utils/utils.go

```diff
@@ -158,36 +158,41 @@ func UDPStoppableRoutine(stopCh <-chan struct{}, name string, decodeFunc decoder
 		payload []byte
 	}
 
-	stopped := atomic.Value{}
-	stopped.Store(false)
 	udpDataCh := make(chan udpData)
+	defer close(udpDataCh)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
 		for {
 			u := udpData{}
 			u.size, u.pktAddr, _ = udpconn.ReadFromUDP(payload)
-			if stopped.Load() == false {
-				if u.size == 0 { // Ignore 0 byte packets.
-					continue
-				}
-				u.payload = make([]byte, u.size)
-				copy(u.payload, payload[0:u.size])
-				udpDataCh <- u
-			} else {
+			if u.size == 0 { // Ignore 0 byte packets.
+				continue
+			}
+			u.payload = make([]byte, u.size)
+			copy(u.payload, payload[0:u.size])
+			select {
+			case <-stopCh:
 				return
+			default:
+				udpDataCh <- u
```

Comment:
Could `udpDataCh <- u` here block forever if:
1/ line 179: enter `default:` case, but not reaching `udpDataCh <- u` yet
2/ line 189:  enter `case <-stopCh:` and `return`
3/ line 180: reached with `udpDataCh <- u` blocking since `udpDataCh` won't be consumed anymore at line 187

?

Reply:
Good catch. Added a drain.
```
	for _ = range udpDataCh {
		// drain
	}
```